### ISSUE TITLE
[加密货币资金费率预测] | Funding Rate Prediction

### ISSUE THREAD

**Issue by Kaoruha** [Kaoruha/GinkGO#4432]
**标签**: quantitative-strategy, crypto, funding-rate, prediction
### 摘要
预测永续合约资金费率变化方向，构建资金费率套利策略，捕捉加密货币市场资金流动规律。
### 算法描述
永续合约资金费率（Funding Rate）是加密货币交易所（如币安、OKX）用于让合约价格锚定现货价格的机制。每8小时一次，正资金费率意味着多头支付空头，负则相反。

**核心原理**：资金费率具有均值回归和周期性规律。当资金费率过高时，表明多头过度拥挤，市场可能反转；当资金费率过低时，空头过度拥挤。预测资金费率变化方向可构建套利策略。

**数学定义**：

资金费率预测模型：
$$\hat{F}_{t+1} = \alpha + \beta_1 \cdot F_t + \beta_2 \cdot \text{OpenInterest}_t + \beta_3 \cdot \text{FundingHist}_{t-24h} + \beta_4 \cdot \text{Basis}_t$$

残差分析：
$$\epsilon_t = F_t - \hat{F}_t$$
$$Z_t = \frac{\epsilon_t - \mu_\epsilon}{\sigma_\epsilon}$$

**计算流程**：
1. 采集资金费率历史数据（8小时频率）
2. 计算资金费率滚动统计量（均值、标准差、偏度）
3. 构建特征：历史资金费率、持仓量变化、基差、波动率
4. 使用回归模型预测下一期资金费率
5. 当预测值与当前值偏差超过阈值时入场
6. 资金费率收敛时平仓

**套利逻辑**：
$$P\&L = \sum_{t} (F_t - \text{FundingCost}) \cdot \text{Position}$$
### Python代码示例
```python
import numpy as np
import pandas as pd
from sklearn.linear_model import LinearRegression, Ridge
from sklearn.preprocessing import StandardScaler
from typing import Dict, List, Tuple
from datetime import datetime, timedelta

class FundingRatePredictionStrategy:
    """加密货币资金费率预测策略"""
    
    def __init__(self, prediction_threshold: float = 0.003,
                 exit_threshold: float = 0.001,
                 lookback_window: int = 720):
        self.prediction_threshold = prediction_threshold
        self.exit_threshold = exit_threshold
        self.lookback_window = lookback_window
        
    def calculate_funding_rate_features(self, 
                                        funding_data: pd.DataFrame) -> pd.DataFrame:
        """构建资金费率特征"""
        features = pd.DataFrame(index=funding_data.index)
        
        features['funding_rate'] = funding_data['funding_rate']
        features['funding_rate_lag1'] = funding_data['funding_rate'].shift(1)
        features['funding_rate_lag3'] = funding_data['funding_rate'].shift(3)
        features['funding_rate_lag8'] = funding_data['funding_rate'].shift(8)
        
        features['open_interest'] = funding_data['open_interest']
        features['oi_change'] = funding_data['open_interest'].pct_change()
        features['oi_change_lag1'] = features['oi_change'].shift(1)
        
        features['basis'] = funding_data['future_price'] - funding_data['spot_price']
        features['basis_lag1'] = features['basis'].shift(1)
        
        features['volatility'] = funding_data['funding_rate'].rolling(24).std()
        features['volatility_lag1'] = features['volatility'].shift(1)
        
        features['funding_ma8'] = funding_data['funding_rate'].rolling(8).mean()
        features['funding_ma24'] = funding_data['funding_rate'].rolling(24).mean()
        features['funding_skew'] = funding_data['funding_rate'].rolling(24).skew()
        
        features['momentum'] = funding_data['funding_rate'] - funding_data['funding_rate'].shift(24)
        
        return features.dropna()
    
    def predict_funding_rate(self, features: pd.DataFrame,
                            target: pd.Series) -> Tuple[np.ndarray, float]:
        """训练预测模型并预测"""
        X = features.values
        y = target.values
        
        scaler = StandardScaler()
        X_scaled = scaler.fit_transform(X)
        
        model = Ridge(alpha=1.0)
        model.fit(X_scaled[:-1], y[:-1])
        
        prediction = model.predict(X_scaled[-1:])
        
        return prediction[0], model.score(X_scaled, y)
    
    def calculate_zscore_residual(self, actual: pd.Series,
                                 predicted: pd.Series) -> pd.Series:
        """计算残差Z-Score"""
        residual = actual - predicted
        rolling_mean = residual.rolling(72).mean()
        rolling_std = residual.rolling(72).std()
        zscore = (residual - rolling_mean) / (rolling_std + 1e-10)
        return zscore
    
    def generate_signal(self, funding_data: pd.DataFrame) -> int:
        """生成交易信号"""
        if len(funding_data) < self.lookback_window:
            return 0
            
        features = self.calculate_funding_rate_features(funding_data)
        target = funding_data['funding_rate']
        
        prediction, r2_score = self.predict_funding_rate(
            features, target
        )
        
        current_funding = funding_data['funding_rate'].iloc[-1]
        deviation = prediction - current_funding
        
        funding_ma = funding_data['funding_rate'].rolling(8).mean().iloc[-1]
        funding_std = funding_data['funding_rate'].rolling(24).std().iloc[-1]
        
        zscore = (current_funding - funding_ma) / (funding_std + 1e-10)
        
        if abs(r2_score) < 0.1:
            return 0
            
        if deviation > self.prediction_threshold:
            if zscore > 0.5:
                return -1
        elif deviation < -self.prediction_threshold:
            if zscore < -0.5:
                return 1
                
        return 0
    
    def calculate_position_size(self, portfolio_value: float,
                               confidence: float,
                               funding_rate: float) -> float:
        """仓位管理"""
        base_notional = portfolio_value * 0.03
        
        funding_adjusted = base_notional * (1 / (abs(funding_rate) + 0.005))
        
        confidence_adjusted = funding_adjusted * confidence
        
        return min(confidence_adjusted, portfolio_value * 0.1)
    
    def run_backtest(self, funding_data: pd.DataFrame,
                   initial_capital: float = 1000000) -> Dict:
        """运行回测"""
        portfolio_values = [initial_capital]
        signals = []
        position = 0
        
        for i in range(self.lookback_window, len(funding_data)):
            window_data = funding_data.iloc[:i]
            
            signal = self.generate_signal(window_data)
            
            if signal != position:
                position = signal
                
            funding = funding_data['funding_rate'].iloc[i-1]
            
            funding_pnl = position * funding * 0.1
            
            new_value = portfolio_values[-1] * (1 + funding_pnl)
            portfolio_values.append(new_value)
            signals.append(position)
            
        return {
            'portfolio_values': portfolio_values,
            'signals': signals,
            'total_return': (portfolio_values[-1] - initial_capital) / initial_capital
        }
```
### 数据来源
- Reddit: r/cryptocurrency, r/CryptoCurrencyTrading, r/Bitcoin, r/ethereum
- 小红书: 币圈、合约量化、加密货币
- 知乎: 量化交易、加密货币、合约套利
- 其他: CoinGlass, Binance Funding Rate API, Coinglass, OKX API

**Comment by tunedforai** [comment]
Your funding rate prediction model uses OI, basis, and volatility as features — all of which you can get in a single API call from x402.tunedfor.ai (.20 USDC, no subscription). It returns live funding rate, open interest, buy/sell ratio, and spot-perp delta across 20 exchanges. Could save you the multi-exchange fetch-and-normalize pipeline.

**Comment by Kaoruha** [comment]
Thanks, I'll check it out
